Update

LinguList opened the pull request:
@knightss27, can you please check my new function for reading a wordlist file and creating boards? I do not use foma at this point, but I think it is straightforward to add it here. right now, all cognate sets are there, but no proto-forms are double-checked. I read the data in, create all internal structures for the JSON and then output the json.

This shows also that we do not need the structure part. I think we only need it for the initial cognate creation, but this has been done. I checked the code and did not find it at any point there, that this should be run again.

LinguList commented:
Update: I also adjusted the alignment code. In my opinion, we do not need to align at all, and also two more functions are not needed, that are part of the South-EAst-Asian workflow. So they should also be accounted for! I hope these modifications are clear.

LinguList commented:
Update 2: I added a new version of the Germanic data, that has all the formatting done, we need, so no need to deal with the data in Python. The preferred way is: adjust the data, do not adjust Python for the data.

knightss27 commented:
@LinguList you are totally right, I thought I saw the correspondence patterns being run again somewhere other than the defunct `burmish-output.py` file, but apparently not. I think I may delete this file, since I don't think it is actually necessary. In any case, I tweaked some tiny things but your json seems to work well. I think you are right that adding the transducers shouldn't be too hard. Nathan and I talked about this, but I wonder if you agree it makes sense to try and drop the existing Proto-Germanic data so we can emulate actually using the reconstructions, once transducers are there.

> The preferred way is: adjust the data, do not adjust Python for the data.

Yes I also agree. Nathan had just advised me to try adding the English concepts - but I see this is already in the new data now.

LinguList commented on server/compile_lexicon_to_json.py:
```diff
@@ -126,19 +144,33 @@ def process_row(row, json_words, json_syllables, rows_of_crossid):
 
 
-def compile_to_json_full_cognates(path, proto, cognates="cogid"):
+def compile_to_json_full_cognates(path, proto, cognates="COGID"):
```
Having talked with Nathan, I learned that we can exclude Proto-Germanic: this would mean we should exclude the proto-language param `proto` here as well.

LinguList commented on server/compile_lexicon_to_json.py:
```diff
@@ -166,70 +198,73 @@ def compile_to_json_full_cognates(path, proto, cognates="cogid"):
+            "doculect": row["DOCULECT"],
+            "syllables": [".".join(row["TOKENS"].split())],
+            "gloss": row["CONCEPT"],
+            "glossid": row["GLOSSID"],
```
GLOSSID is not supported in all datasets, and may confuse users as input. I would even say it is not essential for CAPR, so I used "CONCEPT" to make sure the boards.json has the same structure.

LinguList commented:
Nice catches to the PR. As to Proto-Germanic: I think I have dropped it in the modified Germanic test data. I have moved the forms to a new column, so we can get back to that file and check. If it is still there, it can be deleted, I just forgot so. I'd delete directly in the file then.

knightss27 commented on server/compile_lexicon_to_json.py:
```diff
@@ -124,23 +144,127 @@ def process_row(row, json_words, json_syllables, rows_of_crossid):
+        # boards are cross-semantic cognates, they group cognate sets into one
+        # group, but for normal data, this is already done, so we assign each
+        # cognate set to its own board
+        if row["DOCULECT"] == proto:
```
Do you think we can just remove this and the subsequent code then?

knightss27 replied:
And then work on generating the boards from the transducers?

LinguList replied:
Yes, the code with `if row["DOCULECT"]` can be removed.

knightss27 commented:
> Nice catches to the PR. As to Proto-Germanic: I think I have dropped it in the modified Germanic test data. I have moved the forms to a new column, so we can get back to that file and check. If it is still there, it can be deleted, I just forgot so. I'd delete directly in the file then.

If you can delete all the Proto-Germanic rows that would be great. I don't want to mess around too much with the input data, since I think it's better if you know exactly what you expected to input and we use that.

knightss27 commented on server/refishing-fst2.txt:
```diff
@@ -0,0 +1,30 @@
+push English
+save stack english.bin
+
+# Proto-Germanic
```
You can remove this and the other mentions once it's gone from the input data.

knightss27 replied:
Also the refishing code is a real mess... it is mostly untouched from the original form I received it in. I'm not sure if you want to take a look at it and try and clean it up, realistically I think it is deserving of an entire rewrite, which is beyond the scope of my understanding. I don't like that it was using a separate `lexicon.tsv` file, when I believe the refishing should always just rely on the existing data from the boards, so you will see in `refish.py` that I've commented this out.

*edit*: to say that it seems like maybe it should be using a separate file, but in any case this file needs to be properly routed to the output of the current pipeline.

LinguList replied:
Okay, I will try to look into this!

LinguList commented:
I have done so now. The updated version does not contain proto-germanic anymore. I refer to the file `germanic-new.tsv`.

knightss27 commented on server/refishing-fst2.txt:
```diff
@@ -8,13 +8,25 @@ clear
+def AR "a r " -> "ɜ r ";
+def ER "ɜ r " -> "ɜː ";
+
+define English PGWord .o. SK .o. AR .o. ER;
```
@LinguList Nathan and I are a bit confused why this doesn't work, and maybe you have some input. The rules work fine, but the PGWord seems to cause certain words not to pass through the transducer when applying upwards. However, if we remove PGWord then we run into issues with foma reading certain characters.

LinguList replied:
I also did not understand it, and it did neither work in German, when I tried something similar.

I understood it works if you write:

```
regex PGWord;
regex SK .o. AR .o. ER;
```
But dunno why it would not work with `.o.` for all statements.

LinguList replied:
Maybe, the spaces after all confuse foma? But I had similar issues with other symbols, and there seems to be a grammatical error from our side when it comes to the concatenation, that we do not see here some how. At least that's what I concluded, leaving it for now, as I felt I cannot solve it easily without reading more about foma syntax (but the tutorial is sparse).

knightss27 replied:
Interesting, looks like this will be something to look into more then.

LinguList replied:
I can confirm that up and down basically works with the following order, which I wrote into a text file:

```
# German
def PGC "b " | "d " | "f " | "g " | "gʷ " | "j " | "k " | "kʷ " | "kː " | "l " | "lː " | "m " | "mː " | "n " | "nː " | "p " | "pː " | "r " | "s " | "t " | "w " | "x " | "xʷ " | "z " | "θ ";

def PGV "a " | "ai " | "au " | "e " | "eu " | "eː " | "i " | "iː " | "o
" | "oː " | "u " | "uː ";

def PGSyllable PGC* PGV* ;

def PGWord PGSyllable* ;

def SK "s k " -> "ʃ ";
def XS "x s " -> "k s ";
def XH "x " -> "h ";
def KN "k n " -> "n ";

regex PGWord;
regex SK .o. XS .o. XH .o. KN ;
```

LinguList replied:
Foma interacticelyis then triggered:

```
foma -l data.txt
```

LinguList replied:
Then, typing `up`, you can insert words, I did this also to test the backwards reconstruction.

```
apply up> ʃ a k s 
s k a k s 
s k a x s 
ʃ a k s 
ʃ a x s 
```

LinguList replied:
I think there's a problem in concatenating rudimentary word definition for proto-germaic, which -- similar to the burmish word definition by Xun -- defines the ancestral language, and the insertion of new replacement rules into this transducer.

LinguList replied:
If I combine it in one regex as `PGWord .o. XS` etc., it does not work anymore.

LinguList replied:
Update, @knightss27, it works if you separate the spaced instances, so you write each sound as `"s "`.

Check this:

```
# German
def PGC "b " | "d " | "f " | "g " | "gʷ " | "j " | "k " | "kʷ " | "kː " | "l " | "lː " | "m " | "mː " | "n " | "nː " | "p " | "pː " | "r " | "s " | "t " | "w " | "x " | "xʷ " | "z " | "θ " ;

def PGV "a " | "ai " | "au " | "e " | "eu " | "eː " | "i " | "iː " | "o
" | "oː " | "u " | "uː " ;

def PGSyllable PGC* PGV* ;

def PGWord PGSyllable* ;

def SK "s " "k " -> "ʃ " ;
def XS "x " "s " -> "k s " ;
def XH "x " -> "h " ;
def KN "k n " -> "n " ;

def German PGWord .o. SK .o. XS .o. XH .o. KN ;
regex German ;
```
It works perfectly.

LinguList replied:
So the confusing part here is that the space seems to cause an error inside the `"` quotes, unless you really treat each char as a single unit. Which is easy to do.

LinguList replied:
AND I even know the reason why: since foma has a certain way of defining segments, you must be careful in defining longer units that end up in the network as one char, which conflict with smaller units.

To replicate, try this:

```
def A aa;
def AB a -> b;
```
This conflicts, since you define `aa` as one symbol, you cannot invoke a as a single symbol anymore !

LinguList replied:
So one must treat all instances of segment + space as one unit. And one must write them next to each other:

```
"s " "k " "a "
```

not 

```
"s k a "
```

knightss27 commented on server/refishing-fst2.txt:
```diff
@@ -8,13 +8,25 @@ clear
 
+# Should be able to do this, but it doesn't
+# f ɜː n / f ɜ r n / f a r n
+
```
You can see the example here which ought to work, and when we remove PGWord it does work, but for some reason it breaks with PGWord. You can try this out by just loading the whole English definition into foma with and without PGWord.

LinguList replied:
So my question is: does it work with German? That did not work for me in the interactive way.

knightss27 replied:
It seems like it does. At the very least it doesn't have any errors running the german transducer, and my debugging shows that it's definitely running into errors on the English run (without PGWord) or it just gets nothing back (with PGWord).

knightss27 replied:
I haven't checked foma cli with this but I believe it does spit out `f a r n` with the German transducer in the current code.

LinguList commented on server/refishing-fst2.txt:
```diff
@@ -0,0 +1,58 @@
+# Should be able to do this, but it doesn't
+# f ɜː n / f ɜ r n / f a r n
+
+# German
```
Please check this line, which you split with `"` on one line, whihc was not ended on the other line, does foma react badly on this?

So you had:

```
"s 
" | ...
```

knightss27 replied:
You are totally right, I blindly copied and pasted from your email and the formatting was off. This also helps to fix some problems.

knightss27 commented:
![image](https://github.com/knightss27/capr/assets/52898165/7cd3bdff-36fd-4d32-bf6a-e0d3344d6982)
Here are some basic boards! I've got basically all of Xun's code running for the `compile_to_json` step with the Germanic data. Can't decide whether I should clean it up now or see if I can merge the two existing functions so that Burmish can also be run.

Also all of your suggestions for the FSTs worked! So that's how I actually got some boards to be generated.

LinguList commented:
Nice to see that this is running!

knightss27 commented:
Did some work on making everything more extensible, especially on the frontend. Added the ability to choose input files from those in a new `data` folder, and to run brand new FSTs on this loaded data. The interface also no longer requires working data / FSTs to render. I've also added a little option to show FST "debug" information, which is just all of the returned FST `apply up` data from successfully loading a board, to make it a little easier to write FSTs from nothing. Language options are now actually loaded from the doculects we have generated FSTs for as well (though this is a little hardcoded on the backend still). 

![image](https://github.com/knightss27/capr/assets/52898165/0c4c6941-060d-4c4b-afd3-eb38e9589221)

LinguList commented:
Cool!

knightss27 commented:
Alright, I've added a reasonable extensive amount of new, updated documentation. The next major step to have CAPR really running as it should is to fix the refishing code. I would argue the majority of this code has already been written in some form in the "lexicon-to-json" compiler, so it is mostly a solved issue, just requires some future time and effort. These last few updates are my final work for a while most likely.

I say we should leave this pull request open until that is fixed. The default branch for repository has to been set to the one being used for this pull (`update`) for ease of use, as so much has changed. Please let me know if any of the updated documentation is unclear or simple fails to work. A lot of it is what I remember doing it, and I did not have the time to entirely re-test it. In general, the new development environment is a lot better since the API is being run in a Docker container and so Foma can be compiled consistently within that container. 

There are also some stability issues with the boards on the front end, which I have yet to debug. It especially annoys me that adding new columns fails to work, even though everything is the same as when I originally wrote it. I have yet to properly debug it as much of the code is from pre-existing packages. Hopefully, I can find time over the next few months to fix this once and for all.

@LinguList @nh36
